src/midi/midi_lasynth_model.cpp

```diff
@@ -49,65 +50,91 @@ LASynthModel::LASynthModel(const std::string &rom_name,
 	assert(ctrl_full || (ctrl_a && ctrl_b));
 }
 
-// Checks if its ROMs can be positively found in the provided directory
-bool LASynthModel::InDir(const service_t& service, const std_fs::path& dir) const
+// TODO: Possibly add caching to this function in persistant storage (perhaps a
+// file in ~/.config/dosbox) The identifyROMFile method in the mt32emu library
+// compares with known SHA1 sums so it may be slow for users with lots of files
+// to search. This is mitaged by checking the file size first at least, only
+// running a checksum on files of the correct size.
+std::optional<std_fs::path> LASynthModel::find_rom(const service_t& service,
+                                                   const std_fs::path& dir,
+                                                   const Rom* rom)
 {
-	assert(service);
-
-	auto find_rom = [&service, &dir](const Rom *rom) -> bool {
-		if (!rom)
-			return false;
+	if (!rom) {
+		return {};
+	}
 
-		const auto rom_path = dir / rom->filename;
-		if (!std_fs::exists(rom_path))
-			return false;
+	std::error_code ec;
+	for (const auto& entry : std_fs::directory_iterator(dir, ec)) {
+		if (ec) {
+			continue;
+		}
 
 		mt32emu_rom_info info;
-		if (service->identifyROMFile(&info, rom_path.string().c_str(), nullptr) !=
-		    MT32EMU_RC_OK)
-			return false;
-
-		if (rom->type == ROM_TYPE::UNVERSIONED)
-			return true;
+		if (service->identifyROMFile(&info,
+		                             entry.path().string().c_str(),
+		                             nullptr) != MT32EMU_RC_OK) {
+			continue;
```

Comment:
I think we should log a warning for every single file that could not be identified, so then the user can remove them. If they dumped all sorts of random files into the ROM folder, this sudden influx of warnings will be very educational and will hopefully motivate them to keep their ROM folders clean 😎 🤘🏻